### ISSUE TITLE
Running upgrade to 4.2 from versions below 3.1.32 will give an error with OpenSearch re-indexing

### ISSUE THREAD

**Issue by Euquimides** [craftercms/craftercms#6385]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] The issue is in the develop branch
- [ ] The issue is in the latest released 4.1.x
- [ ] The issue is in the latest released 4.0.x
- [X] The issue is in the latest released 3.1.x

### Describe the issue

If you run an upgrade form versions less than 3.1.32 (for example 3.1.30) to 4.2, will give an error with OpenSearch re-indexing, due to what seems in the logs, to an unknown setting or missing plugin.

### Steps to reproduce

Steps:

1. Follow the steps to upgrade Crafter from 3.1.x to 4.2, as shown in the documentation: https://docs.craftercms.org/en/4.1/by-role/system-admin/upgrade/upgrading-search.html#upgrading-search
2. Pick `yes` when prompted to re-index the site (due to migration from ES to OpenSearch)
3. See the issue

### Relevant log output

```shell
Error in terminal:

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Re-creating Search indexes for sites
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
WARNING: This will delete the current Search site indexes and recreate them.
This is necessary because of a major Search upgrade. Don't proceed 
if you can't have any search downtime.
> Proceed? [(Y)es/(N)o]: y
Caught: upgrade.exceptions.UpgradeException: Error while triggering re-index for 'a1-preview':
upgrade.exceptions.UpgradeException: Error while triggering re-index for 'a1-preview':
	at upgrade.hooks.RecreateIndexesHook.recreateIndex(RecreateIndexesHook.groovy:94)
	at upgrade.hooks.RecreateIndexesHook$_execute_closure1.doCall(RecreateIndexesHook.groovy:52)
	at upgrade.hooks.RecreateIndexesHook.execute(RecreateIndexesHook.groovy:51)
	at upgrade.hooks.PostUpgradeHooks$_execute_closure1.doCall(PostUpgradeHooks.groovy:94)
	at upgrade.hooks.PostUpgradeHooks.execute(PostUpgradeHooks.groovy:93)
	at upgrade.post-upgrade.postUpgrade(post-upgrade.groovy:73)
	at upgrade.post-upgrade.run(post-upgrade.groovy:96)

When checking OpenSearch logs:

[2023-12-19T12:59:28,118][ERROR][o.o.i.i.ManagedIndexCoordinator] [it-Precision-5560] Failed to migrate ISM template
java.lang.IllegalArgumentException: unknown setting [index.lifecycle.name] please check that any required plugins are installed, or check the breaking changes documentation for removed settings
	at org.opensearch.common.settings.AbstractScopedSettings.validate(AbstractScopedSettings.java:608) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.common.settings.AbstractScopedSettings.validate(AbstractScopedSettings.java:549) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.common.settings.AbstractScopedSettings.validate(AbstractScopedSettings.java:519) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.indices.IndicesService.createIndexService(IndicesService.java:789) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.indices.IndicesService.withTempIndexService(IndicesService.java:757) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.action.admin.indices.template.post.TransportSimulateIndexTemplateAction.resolveTemplate(TransportSimulateIndexTemplateAction.java:244) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.action.admin.indices.template.post.TransportSimulateTemplateAction.clusterManagerOperation(TransportSimulateTemplateAction.java:183) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.action.admin.indices.template.post.TransportSimulateTemplateAction.clusterManagerOperation(TransportSimulateTemplateAction.java:70) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.action.support.clustermanager.TransportClusterManagerNodeAction.masterOperation(TransportClusterManagerNodeAction.java:144) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.action.support.clustermanager.TransportClusterManagerNodeAction.clusterManagerOperation(TransportClusterManagerNodeAction.java:153) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.action.support.clustermanager.TransportClusterManagerNodeAction$AsyncSingleAction.lambda$doStart$3(TransportClusterManagerNodeAction.java:271) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.action.ActionRunnable$2.doRun(ActionRunnable.java:88) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.common.util.concurrent.OpenSearchExecutors$DirectExecutorService.execute(OpenSearchExecutors.java:343) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.action.support.clustermanager.TransportClusterManagerNodeAction$AsyncSingleAction.doStart(TransportClusterManagerNodeAction.java:271) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.action.support.clustermanager.TransportClusterManagerNodeAction$AsyncSingleAction.tryAction(TransportClusterManagerNodeAction.java:206) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.action.support.RetryableAction$1.doRun(RetryableAction.java:137) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.common.util.concurrent.OpenSearchExecutors$DirectExecutorService.execute(OpenSearchExecutors.java:343) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.action.support.RetryableAction.run(RetryableAction.java:115) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.action.support.clustermanager.TransportClusterManagerNodeAction.doExecute(TransportClusterManagerNodeAction.java:167) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.action.support.clustermanager.TransportClusterManagerNodeAction.doExecute(TransportClusterManagerNodeAction.java:79) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:218) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.indexmanagement.rollup.actionfilter.FieldCapsFilter.apply(FieldCapsFilter.kt:118) ~[?:?]
	at org.opensearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:216) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.indexmanagement.controlcenter.notification.filter.IndexOperationActionFilter.apply(IndexOperationActionFilter.kt:39) ~[?:?]
	at org.opensearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:216) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.performanceanalyzer.action.PerformanceAnalyzerActionFilter.apply(PerformanceAnalyzerActionFilter.java:78) ~[?:?]
	at org.opensearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:216) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.action.support.TransportAction.execute(TransportAction.java:188) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.action.support.TransportAction.execute(TransportAction.java:107) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.client.node.NodeClient.executeLocally(NodeClient.java:110) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.client.node.NodeClient.doExecute(NodeClient.java:97) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.client.support.AbstractClient.execute(AbstractClient.java:476) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.indexmanagement.indexstatemanagement.migration.ISMTemplateService$simulateTemplate$response$1.invoke(MigrationServices.kt:241) ~[opensearch-index-management-2.9.0.0.jar:2.9.0.0]
	at org.opensearch.indexmanagement.indexstatemanagement.migration.ISMTemplateService$simulateTemplate$response$1.invoke(MigrationServices.kt:241) ~[opensearch-index-management-2.9.0.0.jar:2.9.0.0]
	at org.opensearch.indexmanagement.opensearchapi.OpenSearchExtensionsKt.suspendUntil(OpenSearchExtensions.kt:217) ~[opensearch-index-management-2.9.0.0.jar:2.9.0.0]
	at org.opensearch.indexmanagement.indexstatemanagement.migration.ISMTemplateService.simulateTemplate(MigrationServices.kt:241) ~[opensearch-index-management-2.9.0.0.jar:2.9.0.0]
	at org.opensearch.indexmanagement.indexstatemanagement.migration.ISMTemplateService.getIndexTemplates(MigrationServices.kt:141) ~[opensearch-index-management-2.9.0.0.jar:2.9.0.0]
	at org.opensearch.indexmanagement.indexstatemanagement.migration.ISMTemplateService.doMigration$suspendImpl(MigrationServices.kt:91) ~[opensearch-index-management-2.9.0.0.jar:2.9.0.0]
	at org.opensearch.indexmanagement.indexstatemanagement.migration.ISMTemplateService.doMigration(MigrationServices.kt) ~[opensearch-index-management-2.9.0.0.jar:2.9.0.0]
	at org.opensearch.indexmanagement.indexstatemanagement.ManagedIndexCoordinator$initTemplateMigration$scheduledJob$1$1.invokeSuspend(ManagedIndexCoordinator.kt:567) [opensearch-index-management-2.9.0.0.jar:2.9.0.0]
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33) [kotlin-stdlib-1.8.21.jar:1.8.21-release-380(1.8.21)]
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56) [kotlinx-coroutines-core-jvm-1.3.9.jar:?]
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571) [kotlinx-coroutines-core-jvm-1.3.9.jar:?]
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:738) [kotlinx-coroutines-core-jvm-1.3.9.jar:?]
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678) [kotlinx-coroutines-core-jvm-1.3.9.jar:?]
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665) [kotlinx-coroutines-core-jvm-1.3.9.jar:?]
```

**Comment by sumerjabri** [comment]
@russdanner let's check who's left in 3.1 and decide if we should prioritize this.

**Comment by phuongnq** [comment]
@sumerjabri @Euquimides I couldn't reproduce this issue as of current but there are some notes and a fix (due to another issue) for the UM. 

Here is the step to upgrade from 3.1.x to 4.2.0 (I tested with 3.1.30). There are some change compares to our doc since in 4.2.0, we use Java 21 and MariaDB 11.

* Set Java to 17 (This is required to run the upgrade-search.sh script since the ES 7 won't run with Java 21)

* Run the upgrade seach script
```
./upgrade-search.sh /home/pnguyen/Downloads/crafter-cms-authoring-3.1.30/crafter --stay-alive
```

A sample log: 
[upgrade_search.log](https://github.com/user-attachments/files/15802890/upgrade_search.log)

* Set Java to 21

* Upgrade target **without** backup the `bin` folder (Since the MariaDB version is now 11 for 4.2.0, we cannot start the old MariaDB from the new bundle. Please manually backup it beforehand)

```
./upgrade-target.sh /home/pnguyen/Downloads/crafter-cms-authoring-3.1.30/crafter
```

* Before performing post-upgrade, please upgrade the DB

```
cd ~/Downloads/crafter-cms-authoring-3.1.30/crafter
./bin/crafter.sh upgradedb
```
A sample log: 
[upgrade_db.log](https://github.com/user-attachments/files/15802895/upgrade_db.log)

* Run post-upgrade script

```
/home/pnguyen/Downloads/crafter-cms-authoring-3.1.30/crafter/bin/upgrade/post-upgrade.sh
````

A sample log: 
[post_upgrade.log](https://github.com/user-attachments/files/15802904/post_upgrade.log)

**Comment by phuongnq** [comment]
@sumerjabri Regarding the PR, it fixed an issue that when upgrading from 3.1.x to 4.2.0, the `refresh_token` and `access_token` tables require the field `last_updated_on` to NOT NULL. This dues to the following UM script:
https://github.com/craftercms/studio/blob/develop/src/main/resources/crafter/studio/database/upgrade-3.1.x-to-4.0.0.32.sql#L138

```
CREATE TABLE IF NOT EXISTS `refresh_token` (
  `user_id` BIGINT(20) NULL DEFAULT NULL,
  `token` VARCHAR(50) NOT NULL,
  `last_updated_on` TIMESTAMP NOT NULL,
  `created_on` TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
  PRIMARY KEY (`user_id`))
```

This does not align with the createDLL.sql where we allow the column to be NULL:
https://github.com/craftercms/studio/blob/develop/src/main/resources/crafter/studio/database/createDDL.sql#L609
```
CREATE TABLE IF NOT EXISTS `refresh_token`
(
    `user_id` BIGINT(20) PRIMARY KEY, -- Add FK
    `token` VARCHAR(50) NOT NULL,
    `last_updated_on` TIMESTAMP,
    `created_on` TIMESTAMP DEFAULT CURRENT_TIMESTAMP
)
```

This causes an issue with inserting rows to the tables because we don't specify the `last_updated_on` in the insert query. The PR fixes this to remove the required `NOT NULL`.

**Comment by jmendeza** [comment]
@phuongnq is this not an issue when upgrading to 4.1.x?

**Comment by sumerjabri** [comment]
@alhambrav please add to the documentation steps for upgrades. See @phuongnq 's excellent steps above.

**Comment by phuongnq** [comment]
> @phuongnq is this not an issue when upgrading to 4.1.x?

@jmendeza We don't have this issue in 4.1.x. Could be due to the MariaDB version. But anyway, I don't think we use the field `last_updated_on` anywhere in the code.

**Comment by Euquimides** [comment]
Hi! Updated and rechecked this issue. Fix verified for Crafter v4.2.0. Thank you!

**Comment by alhambrav** [comment]
Assigning the ticket to @yacdaniel to review the docs PR https://github.com/craftercms/docs/pull/1843
